### DISCUSSION TITLE
django-recaptcha in wizard forms

### DISCUSSION THREAD

**Discussion by barseghyanartur** [django-recaptcha/django-recaptcha#263] [Q&A] [answered]
At the moment when used in form wizards, captcha becomes invalid on the last step. Any solution to this?

**Comment by adamchainz** [comment]
Can you expand a bit more - is the captcha in the last step, or is it in an earlier step and then counts as invalid later? I can see the problem in the latter case, as presumably validation runs again when the whole form wizard is done and the captcha value can only be validated once.

**Comment by pringelmann** [comment]
This will be a problem regardless of what step you place the captcha on. (as the validation will run again at the end of the wizard anyway) 
A simple workaround is to isolate your captcha to its own view/form.

**Comment by adamchainz** [comment]
Isn't there something the field could do to work with wizards by caching the 'passed' state of the captcha after the first validation run?

**Comment by pringelmann** [comment]
Good suggestion. So it just locks to a validated state? Will look into it :)

**Comment by realmhamdy** [comment]
I'm using [django-formtools](https://django-formtools.readthedocs.io/en/latest/) for a wizard and struggled with the same issue. I found the `sys._getframe()` hack in `ReCaptchaField.get_remote_ip()` method and could use it to let the `ReCaptchaField` know when it's a relvalidation.
First create a ReCaptchaField subclass:
```
import sys
from captcha.fields import ReCaptchaField


class ValidateOnceReCaptchaField(ReCaptchaField):

    def clean(self, values):
        # find the 'revalidating' value in stack
        frame = sys._getframe()
        max_depth = 25
        while frame and max_depth != 0:
            if 'revalid' in frame.f_locals:
                return values[0]
            max_depth -= 1
            frame = frame.f_back
        return super(ValidateOnceReCaptchaField, self).clean(values)
```
And in your `WizardView` subclass, override `render_done`:
```
class MyView(WizardView):
  
    def render_done(self, form, **kwargs):
        revalid = True
        return super(MyView, self).render_done(form, **kwargs)
```

This way only the ReCaptcha revalidation is skipped. If someone knows a better way to pass values from a view to a form field please post it here.

**Comment by richtier** [comment, marked as answer]
we solved the problem in a different way that does not use the `sys._getframe()` hack.

https://github.com/uktrade/directory-ui-supplier/pull/337/files
